rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -0,0 +1,248 @@
+import typing
+from typing import List, Optional, Text, Dict, Tuple, Union
+import copy
+import numpy as np
+from collections import defaultdict, OrderedDict
+import scipy.sparse
+
+from rasa.utils.tensorflow.model_data import Data
+from rasa.utils.tensorflow.constants import SEQUENCE
+
+if typing.TYPE_CHECKING:
+    from rasa.utils.features import Features
+
+MASK = "mask"
+
+
+def surface_attributes(
+    tracker_state_features: List[List[Dict[Text, List["Features"]]]]
+) -> Dict[Text, List[List[List["Features"]]]]:
+    """Restructure the input.
+
+    Args:
+        tracker_state_features: a dictionary of attributes (INTENT, TEXT, ACTION_NAME,
+            ACTION_TEXT, ENTITIES, SLOTS, FORM) to a list of features for all
+            dialogue turns in all training trackers
+
+    Returns:
+        A dictionary of attributes to a list of features for all dialogue turns
+        and all training trackers.
+    """
+    # collect all attributes
+    attributes = set(
+        attribute
+        for features_in_tracker in tracker_state_features
+        for features_in_turn in features_in_tracker
+        for attribute in features_in_turn.keys()
+    )
+
+    attribute_to_features = defaultdict(list)
+
+    for features_in_tracker in tracker_state_features:
+        intermediate_features = defaultdict(list)
+
+        for features_in_dialogue in features_in_tracker:
+            for attribute in attributes:
+                # if attribute is not present in the example, populate it with None
+                intermediate_features[attribute].append(
+                    features_in_dialogue.get(attribute)
+                )
+
+        for key, value in intermediate_features.items():
+            attribute_to_features[key].append(value)
+
+    return attribute_to_features
+
+
+def create_zero_features(
+    tracker_features: List[List[List["Features"]]],
+) -> List["Features"]:
+    # all features should have the same types
+    """
+    Computes default feature values for an attribute;
+    Args:
+        tracker_features: list containing all feature values encountered
+        in the dataset for an attribute;
+    """
+
+    example_features = next(
+        iter(
+            [
+                list_of_features
+                for turn_features in tracker_features
+                for list_of_features in turn_features
+                if list_of_features is not None
+            ]
+        )
+    )
+
+    # create zero_features for nones
+    zero_features = []
+    for features in example_features:
+        new_features = copy.deepcopy(features)
+        if features.is_dense():
+            new_features.features = np.zeros_like(features.features)
+        if features.is_sparse():
+            new_features.features = scipy.sparse.coo_matrix(
+                features.features.shape, features.features.dtype
+            )
+        zero_features.append(new_features)
+
+    return zero_features
+
+
+def convert_to_data_format(

```

Comment:
this one should be split into smaller functions

Reply:
@tabergma could you please take this over

Reply:
yes, will take care of this